yarn/src/main/scala/org/apache/spark/deploy/yarn/ContainerPlacementStrategy.scala

```diff
@@ -0,0 +1,206 @@
+ *
+ * 5. If containers are existed and existing localities can fully cover the requested localities.
+ * For example if we have 5 containers on each node (host1: 5, host2: 5, host3: 5, host4: 5),
+ * which could cover the current requested localities. This algorithm will allocate all the
+ * requested containers with no localities.
+ */
+private[yarn] class LocalityPreferredContainerPlacementStrategy(
+    val sparkConf: SparkConf,
+    val yarnConf: Configuration,
+    val yarnAllocator: YarnAllocator) extends ContainerPlacementStrategy {
+
+  // Number of CPUs per task
+  private val CPUS_PER_TASK = sparkConf.getInt("spark.task.cpus", 1)
+
+  // Get the required cores of locality aware task
+  private def localityAwareTaskCores(localityAwarePendingTasks: Int): Int = {
+    localityAwarePendingTasks * CPUS_PER_TASK
+  }
+
+  // Get the expected number of locality aware containers
+  private def expectedLocalityAwareContainerNum(localityAwarePendingTasks: Int): Int = {
+    (localityAwareTaskCores(localityAwarePendingTasks) +
+      yarnAllocator.resource.getVirtualCores - 1) /
+        yarnAllocator.resource.getVirtualCores
+  }
+
+  // Update the expected locality distribution by considering the existing allocated container
+  // host distributions.
+  private def updateExpectedLocalityToCounts(localityAwarePendingTasks: Int,
+      preferredLocalityToCounts: Map[String, Int]): Map[String, Int] = {
+    val totalPreferredLocalities = preferredLocalityToCounts.values.sum
+    preferredLocalityToCounts.map { case (host, count) =>
+      val expectedCount =
+        count.toDouble * expectedLocalityAwareContainerNum(localityAwarePendingTasks) /
+          totalPreferredLocalities
+      val existedCount = yarnAllocator.allocatedHostToContainersMap.get(host)
+        .map(_.size)
+        .getOrElse(0)
+
+      if (expectedCount > existedCount) {
+        // Get the actual container number if existing container can not fully satisfy the
+        // expected number of container
+        (host, (expectedCount - existedCount).ceil.toInt)
+      } else {
+        // If the current existed container number can fully satisfy the expected number of
+        // containers, set the required containers to be 0
+        (host, 0)
+      }
+    }
+  }
+
+  def localityOfRequestedContainers(
+      numContainer: Int,
+      numLocalityAwarePendingTasks: Int,
+      preferredLocalityToCounts: Map[String, Int]
+    ): Array[ContainerLocalityPreferences] = {
+    val updatedLocalityToCounts =
+      updateExpectedLocalityToCounts(numLocalityAwarePendingTasks, preferredLocalityToCounts)
+    val updatedLocalityAwareContainerNum = updatedLocalityToCounts.values.sum
+
+    // The number of containers to allocate, divided into two groups, one with preferred locality,
+    // and the other without locality preference.
+    var requiredLocalityFreeContainerNum: Int = 0
+    var requiredLocalityAwareContainerNum: Int = 0
+
```

Comment:
Can this if/else block be replaced with:

    requiredLocalityFreeContainerNum = math.max(0, numContainer - updatedLocalityAwareContainerNum)
    requiredLocalityAwareContainerNum = numContainer - requiredLocalityFreeContainerNum

?